java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java

```diff
@@ -0,0 +1,193 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import org.sonar.java.checks.methods.MethodInvocationMatcher;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.InstanceOfTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.ParenthesizedTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S2097",
+  name = "\"equals(Object obj)\" should test argument type",
+  tags = {"bug"},
+  priority = Priority.BLOCKER)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("5min")
+public class EqualsArgumentTypeCheck extends SubscriptionBaseVisitor {
+
+  private static final MethodInvocationMatcher EQUALS_MATCHER = MethodInvocationMatcher.create()
+    .name("equals")
+    .addParameter("java.lang.Object");
+
+  private boolean typeChecked;
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (hasSemantic()) {
+      MethodTree methodTree = (MethodTree) tree;
+      if (methodTree.block() != null && "equals".equals(methodTree.symbol().name()) && methodTree.parameters().size() == 1) {
+        Symbol parameterSymbol = methodTree.parameters().get(0).symbol();
+        if (parameterSymbol.type().is("java.lang.Object")) {
+          typeChecked = false;
+          methodTree.accept(new EqualsMethodVisitor(parameterSymbol));
+          if (!typeChecked) {
```

Comment:
I'm not fond of having the local variable `typeChecked ` modified in the `ExpressionVisitor`, especially when we used an `EqualsMethodVisitor` to trigger it. I would rather prefer asking the visitor, after its visit, if it encountered a type check. it would allow the two visitor to be `static`, but I agree that it will also imply a bit more of code on visitors side.

Moreover, why not using always the same `ExpressionVisitor`?